### PR TITLE
DC: Fixed behavior of `predict` with `output_type=probability`

### PR DESCRIPTION
**Pull request by shantanuchhabra**
Thanks to @nickjong for pointing out that our implementation of `output_type="probability"` has been wrong all this while in `predict`.

The `output_type="probability"` option is exclusively meant for binary classifiers. This option is meant to provide the probability of the True class in binary classifiers.
For example, if one is building a binary classifier for "cat" vs "not cat", output_type="probability" should output the probability of the data point being "cat", not the probability of the predicted class.

This PR is an effort to fix this behavior (that we had misunderstood until now), and also adds a C++ unit test.

### CODE REVIEW

**Comment by Jarvi-Izana** [comment]
I don't have the full context of the behavior of 5.8. I guess you still want to support `probability_vector` when the number of classification for more than 2 classes.

**Comment by shantanuchhabra** [comment]
> I don't have the full context of the behavior of 5.8. I guess you still want to support `probability_vector` when the number of classification for more than 2 classes.

Yes, we still support `probability_vector` as an output type. We first check to make sure the output type provided by the user is one of "class", "probability", and "probability_vector". The output from `perform_inference` is already aligned with what we want to output for `probability_vector` so we only check for `class` and `probability` and modify the result accordingly. Does that make sense?

**Comment by Jarvi-Izana** [comment]
> > I don't have the full context of the behavior of 5.8. I guess you still want to support `probability_vector` when the number of classification for more than 2 classes.
> 
> Yes, we still support `probability_vector` as an output type. We first check to make sure the output type provided by the user is one of "class", "probability", and "probability_vector". The output from `perform_inference` is already aligned with what we want to output for `probability_vector` so we only check for `class` and `probability` and modify the result accordingly. Does that make sense?

makes sense.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 1]
```diff
@@ -700,14 +700,29 @@ gl_sarray drawing_classifier::predict(gl_sframe data, std::string output_type) {
...
+    if (num_classes > 2) {
+      log_and_throw(
+          "Use probability_vector in case of multi-class classification.");
+    }
+
+    auto true_class_probability = [=](const flexible_type& ft) {
       const flex_vec& prob_vec = ft.get<flex_vec>();
-      auto max_it = std::max_element(prob_vec.begin(), prob_vec.end());
-      return *max_it;
+      size_t last_index = num_classes - 1;  // either 0 or 1
```
Comment (Jarvi-Izana) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: Could you explain why only the last probability is returned?

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 2]
```diff
@@ -700,14 +700,29 @@ gl_sarray drawing_classifier::predict(gl_sframe data, std::string output_type) {
...
+      return *(prob_vec.begin() + last_index);
```
Comment (nickjong) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: Could we just return `prob_vec.back()` here?

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 1 above)
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: With True and False labels, i.e. "cat" and "not cat", we assume that the False label has index 0 and the True label has index 1. I'm not sure where we broadcast this to users (probably somewhere).

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 2 above)
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: Done, thanks!

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 3]
```diff
@@ -700,14 +700,28 @@ gl_sarray drawing_classifier::predict(gl_sframe data, std::string output_type) {
...
+
+    size_t num_classes = read_state<flex_int>("num_classes");
+    DASSERT_GT(num_classes, 0);
+
+    if (num_classes > 2) {
+      log_and_throw(
+          "Use probability_vector in case of multi-class classification.");
+    }
+
+    auto true_class_probability = [=](const flexible_type& ft) {
```
Comment (nickjong) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: Leave an empty capture list when not capturing anything with a lambda expression

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 1 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: The relative order of `cat` and `not-cat` should depend on the label from the training data. set, i.e, data iterator may return label vector of <cat, not-cat> or <cat, not-cat>. You may. check where `num_classes` is introduced in `init_training`.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 4]
```diff
@@ -700,14 +700,28 @@ gl_sarray drawing_classifier::predict(gl_sframe data, std::string output_type) {
...
       const flex_vec& prob_vec = ft.get<flex_vec>();
-      auto max_it = std::max_element(prob_vec.begin(), prob_vec.end());
-      return *max_it;
+      return prob_vec.back();
```
Comment (Jarvi-Izana) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: I think we need to align with the label order generated from the training dataset.

Comment (nickjong) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: Could you sync with @hoytak? We should align with what all the supervised-learning toolkits do.

Comment (srikris) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: This should align with the label order in the training set that are sorted according to class name. So for example:

```
model.classes -> ['cat', 'dog', 'foosa']
```
and predictions should be in that order
```
[p_cat, p_dog, p_foosa
```

Sadly the binary classifier is a little annoying because all other packages have the convention of using a "reference" class to minimize multi-collinearity and have fewer parameters while this isn't much of a concept in deep learning based models.

Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: @hoytak , @Jarvi-Izana , and I discussed this at length and here are our thoughts:

1. For `output_type="probability"` in a binary classifier setting, all our supervised learning toolkits currently follow the approach of sorting labels lexicographically, and returning the probability of the last label.
2. This approach is kind of arbitrary, since the labels could be "cat" and "not cat", in which case the user will end up with probabilities for the label, "not cat".
3. After some discussion, we came to the conclusion that for TC 6.0, we should modify this behavior across Drawing Classifier, Logistic Classifier, SVMs, and other supervised learning toolkits to the following: `output_type="probability"` should only apply if the target values in the data are `0` and `1`. This issue is tracked in #2667 .

Here is what I suggest w.r.t this PR:

1. Let's merge this PR in its current state where it matches what we currently have for all supervised learning toolkits.
2. Let's address #2667 before TC 6.0 and make a code change in this logic (and elsewhere) then.

@nickjong @srikris @Jarvi-Izana @hoytak Thoughts?

Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: As long as it's book-kept, I'm good.

Comment (nickjong) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: I don't think we have the time to adequately design, develop, and test this kind of change for 6.0.
